### PR TITLE
chore: bump Node.js to v16.2.0

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Upgrading Node.js to v16.2.0. A work in progress, which ideally will go out in Electron v14.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: : Updated Node.js to v16.2.0.

### CODE REVIEW

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch [hunk 1]
```diff
@@ -8,86 +8,288 @@ and https://github.com/nodejs/node/pull/31178 that is not currently compatible
...
+diff --git a/src/node.cc b/src/node.cc
+index a99f2f4f01b16fc70f3038a25da22c788714b3f6..ceb532648b33a6c2f3b8a135b315985cfff5419e 100644
+--- a/src/node.cc
++++ b/src/node.cc
+@@ -1020,7 +1020,7 @@ InitializationResult InitializeOncePerProcess(int argc, char** argv) {
+     return result;
+   }
+ 
+-#if HAVE_OPENSSL
++#if HAVE_OPENSSL && !defined(OPENSSL_IS_BORINGSSL)
```
Comment (nornagon) [inline patches/node/fix_comment_out_incompatible_crypto_modules.patch:279]: Did we support this before?

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch [hunk 2]
```diff
@@ -8,86 +8,288 @@ and https://github.com/nodejs/node/pull/31178 that is not currently compatible
...
+diff --git a/src/node_metadata.h b/src/node_metadata.h
+index 4486d5af2c1622c7c8f44401dc3ebb986d8e3c2e..db1769f1b3f1617ed8dbbea57b5e324183b42be2 100644
+--- a/src/node_metadata.h
++++ b/src/node_metadata.h
+@@ -6,7 +6,7 @@
+ #include <string>
+ #include "node_version.h"
+ 
+-#if HAVE_OPENSSL
++#if 0
```
Comment (nornagon) [inline patches/node/fix_comment_out_incompatible_crypto_modules.patch:292]: what does this fix?

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 3]
```diff
@@ -135,8 +135,20 @@ index 863907bafd81920c40bccaf89299a2bcdb1be79e..c6f1c146586d0af096def5d6ee734373
...
      algo: 'chacha20-poly1305',
+@@ -669,7 +697,10 @@ for (const test of TEST_CASES) {
+ 
+ {
+   // CCM cipher without data should not crash, see https://github.com/nodejs/node/issues/38035.
+-  const algo = 'aes-128-ccm';
++  common.printSkipMessage(`unsupported aes-128-ccm test`);
++  return;
++
++  const algo = 'aes-192-cbc';
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:146]: if we don't run the test there's no need for this updated `const` right?

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 4]
```diff
@@ -866,8 +913,82 @@ index 2d005c89db3f09657469a73dca2d28eecae756e2..e2da39f3f87c557db43aa33cf3505e14
...
+index c85a79b4854369e35fbe89833e9df9a12065671e..48fb14bd044c87ae5bdbe2563f70d42084f24d09 100644
+--- a/test/parallel/test-crypto-x509.js
++++ b/test/parallel/test-crypto-x509.js
+@@ -104,7 +104,9 @@ const der = Buffer.from(
+     '84:AC:5B:08:9A:20:89:B6:8F:D6'
+   );
+   assert.strictEqual(x509.keyUsage, undefined);
+-  assert.strictEqual(x509.serialNumber, 'ECC9B856270DA9A8');
++
++  const serialPattern = /ECC9B856270DA9A8|ecc9b856270da9a8/;
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:926]:
```suggestion
+  const serialPattern = /ECC9B856270DA9A8/i;
```

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 5]
```diff
@@ -866,8 +913,82 @@ index 2d005c89db3f09657469a73dca2d28eecae756e2..e2da39f3f87c557db43aa33cf3505e14
...
++  assert.strictEqual(serialPattern.test(x509.serialNumber), true);
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:927]:
```suggestion
+  assert.match(x509.serialNumber, serialPattern);
```

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 6]
```diff
@@ -866,8 +913,82 @@ index 2d005c89db3f09657469a73dca2d28eecae756e2..e2da39f3f87c557db43aa33cf3505e14
...
++                      '6FBD809CB8E037A98B71708E001242E4DEB54C6164885F599DD87' +
++                      'A23215745955BE20417E33C4D0D1B80C9DA3DE419A2607195D2FB' +
++                      '75657B0BBFB5EB7D0BBA5122D1B6964C7B570D50B8EC001EEB68D' +
++                      'FB584437508F3129928D673B30A3E0BF4F50609E6371';
++
++  const modulusBSSL = 'ef5440701637e28abb038e5641f828d834c342a9d25edbb86a2bf' +
++                      '6fbd809cb8e037a98b71708e001242e4deb54c6164885f599dd87' +
++                      'a23215745955be20417e33c4d0d1b80c9da3de419a2607195d2fb' +
++                      '75657b0bbfb5eb7d0bba5122d1b6964c7b570d50b8ec001eeb68d'+
++                      'fb584437508f3129928d673b30a3e0bf4f50609e6371';
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:945]: `.toLowerCase()`?

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 7]
```diff
@@ -866,8 +913,82 @@ index 2d005c89db3f09657469a73dca2d28eecae756e2..e2da39f3f87c557db43aa33cf3505e14
...
+@@ -213,11 +227,7 @@ const der = Buffer.from(
+         'CA Issuers - URI:http://ca.nodejs.org/ca.cert' :
+         'OCSP - URI:http://ocsp.nodejs.org/\n' +
+         'CA Issuers - URI:http://ca.nodejs.org/ca.cert\n',
+-    modulus: 'EF5440701637E28ABB038E5641F828D834C342A9D25EDBB86A2BF' +
+-             '6FBD809CB8E037A98B71708E001242E4DEB54C6164885F599DD87' +
+-             'A23215745955BE20417E33C4D0D1B80C9DA3DE419A2607195D2FB' +
+-             '75657B0BBFB5EB7D0BBA5122D1B6964C7B570D50B8EC001EEB68D' +
+-             'FB584437508F3129928D673B30A3E0BF4F50609E6371',
++    modulusPattern: new RegExp(`${modulusOSSL}|${modulusBSSL}`),
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:959]:
```suggestion
+    modulusPattern: new RegExp(modulusOSSL, 'i'),
```

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 8]
```diff
@@ -866,8 +913,82 @@ index 2d005c89db3f09657469a73dca2d28eecae756e2..e2da39f3f87c557db43aa33cf3505e14
...
+     bits: 1024,
+     exponent: '0x10001',
+     valid_from: 'Nov 16 18:42:21 2018 GMT',
+@@ -226,7 +236,7 @@ const der = Buffer.from(
+     fingerprint256:
+       'B0:BE:46:49:B8:29:63:E0:6F:63:C8:8A:57:9C:3F:9B:72:' +
+       'C6:F5:89:E3:0D:84:AC:5B:08:9A:20:89:B6:8F:D6',
+-    serialNumber: 'ECC9B856270DA9A8'
++    serialNumberPattern: /ECC9B856270DA9A8|ecc9b856270da9a8/
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:968]:
```suggestion
+    serialNumberPattern: /ECC9B856270DA9A8/i
```

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 9]
```diff
@@ -866,8 +913,82 @@ index 2d005c89db3f09657469a73dca2d28eecae756e2..e2da39f3f87c557db43aa33cf3505e14
...
+   };
+ 
+   const legacyObject = x509.toLegacyObject();
+@@ -235,7 +245,7 @@ const der = Buffer.from(
+   assert.strictEqual(legacyObject.subject, legacyObjectCheck.subject);
+   assert.strictEqual(legacyObject.issuer, legacyObjectCheck.issuer);
+   assert.strictEqual(legacyObject.infoAccess, legacyObjectCheck.infoAccess);
+-  assert.strictEqual(legacyObject.modulus, legacyObjectCheck.modulus);
++  assert.strictEqual(legacyObjectCheck.modulusPattern.test(legacyObject.modulus), true);
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:977]:
```suggestion
+  assert.match(legacyObject.modulus, legacyObjectCheck.modulusPattern);
```

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 10]
```diff
@@ -866,8 +913,82 @@ index 2d005c89db3f09657469a73dca2d28eecae756e2..e2da39f3f87c557db43aa33cf3505e14
...
+   assert.strictEqual(legacyObject.exponent, legacyObjectCheck.exponent);
+   assert.strictEqual(legacyObject.valid_from, legacyObjectCheck.valid_from);
+@@ -244,7 +254,5 @@ const der = Buffer.from(
+   assert.strictEqual(
+     legacyObject.fingerprint256,
+     legacyObjectCheck.fingerprint256);
+-  assert.strictEqual(
+-    legacyObject.serialNumber,
+-    legacyObjectCheck.serialNumber);
++  assert.strictEqual(legacyObjectCheck.serialNumberPattern.test(legacyObject.serialNumber), true);
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:988]:
```suggestion
+  assert.match(legacyObject.serialNumber, legacyObjectCheck.serialNumberPattern);
```

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 11]
```diff
@@ -931,16 +1052,77 @@ index 6b72dbd21cd07dcd0c6fc01d36195b33aeb295a2..1729094574d1ee0321ad0118e3cf97d3
...
+ 
+ // Create TLS1.2 server
+diff --git a/test/parallel/test-tls-getcertificate-x509.js b/test/parallel/test-tls-getcertificate-x509.js
+index 5be788f67931131256f7fb0ab802cb0edee58173..13963fbeee1bbe5d4a25f1ef8bb58c7a0ad1eb23 100644
+--- a/test/parallel/test-tls-getcertificate-x509.js
++++ b/test/parallel/test-tls-getcertificate-x509.js
+@@ -21,8 +21,8 @@ server.once('secureConnection', common.mustCall(function(socket) {
+   const cert = socket.getX509Certificate();
+   assert(cert instanceof X509Certificate);
+   assert.strictEqual(
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:1111]: `assert.match`

File: patches/node/fix_crypto_tests_to_run_with_bssl.patch [hunk 12]
```diff
@@ -931,16 +1052,77 @@ index 6b72dbd21cd07dcd0c6fc01d36195b33aeb295a2..1729094574d1ee0321ad0118e3cf97d3
...
+-    'D0082F458B6EFBE8');
++    /d0082f458b6efbe8|D0082F458B6EFBE8/.test(cert.serialNumber), true
++  )
+ }));
+ 
+ server.listen(0, common.mustCall(function() {
+@@ -34,8 +34,7 @@ server.listen(0, common.mustCall(function() {
+     assert(peerCert.issuerCertificate instanceof X509Certificate);
+     assert.strictEqual(peerCert.issuerCertificate.issuerCertificate, undefined);
+     assert.strictEqual(
```
Comment (nornagon) [inline patches/node/fix_crypto_tests_to_run_with_bssl.patch:1122]: `assert.match`

File: patches/node/fix_key_gen_apis_are_not_available_in_boringssl.patch [hunk 13]
```diff
@@ -5,72 +5,20 @@ Subject: fix: key gen APIs are not available in BoringSSL
...
+diff --git a/src/crypto/crypto_common.cc b/src/crypto/crypto_common.cc
+index f4b7bd3ad8548a0b69943ddea669e6f1991b7a49..221d652fa7de246e5f69fcf392e334087bac0199 100644
+--- a/src/crypto/crypto_common.cc
++++ b/src/crypto/crypto_common.cc
+@@ -242,7 +242,7 @@ const char* GetClientHelloALPN(const SSLPointer& ssl) {
    const unsigned char* buf;
    size_t len;
    size_t rem;
 -
++#ifndef OPENSSL_IS_BORINGSSL
```
Comment (nornagon) [inline patches/node/fix_key_gen_apis_are_not_available_in_boringssl.patch:17]: How come this `#ifndef` is moved to after the variables? Doesn't that make the variables warn as unused...?

File: patches/node/fix_key_gen_apis_are_not_available_in_boringssl.patch [hunk 14]
```diff
@@ -79,15 +27,15 @@ index 6473b652ac95609aff555d99be38b48a5aa513a5..caaaf19dc02101c2024b511780c94fc8
  }
  
  const char* GetClientHelloServerName(const SSLPointer& ssl) {
-+#ifndef OPENSSL_IS_BORINGSSL
    const unsigned char* buf;
    size_t len;
    size_t rem;
 -
++#ifndef OPENSSL_IS_BORINGSSL
```
Comment (nornagon) [inline patches/node/fix_key_gen_apis_are_not_available_in_boringssl.patch:34]: here too

File: patches/node/fix_key_gen_apis_are_not_available_in_boringssl.patch [hunk 15]
```diff
@@ -124,3 +72,54 @@ index 6473b652ac95609aff555d99be38b48a5aa513a5..caaaf19dc02101c2024b511780c94fc8
...
+   if (!param_ctx ||
+       EVP_PKEY_paramgen_init(param_ctx.get()) <= 0 ||
+       EVP_PKEY_CTX_set_dsa_paramgen_bits(
+@@ -49,7 +49,7 @@ EVPKeyCtxPointer DsaKeyGenTraits::Setup(DsaKeyPairGenConfig* params) {
+       return EVPKeyCtxPointer();
+     }
+   }
+-
++#endif
```
Comment (nornagon) [inline patches/node/fix_key_gen_apis_are_not_available_in_boringssl.patch:93]: are we correctly signalling failure here?

File: patches/node/fix_use_crypto_impls_for_compat.patch [hunk 16]
```diff
@@ -16,20 +16,24 @@ We can revisit this once that happens and determine a more mutually
...
+@@ -110,7 +110,7 @@ unsigned int GetBytesOfRS(const ManagedEVPPKey& pkey) {
    if (base_id == EVP_PKEY_DSA) {
-     DSA* dsa_key = EVP_PKEY_get0_DSA(pkey.get());
+     const DSA* dsa_key = EVP_PKEY_get0_DSA(pkey.get());
      // Both r and s are computed mod q, so their width is limited by that of q.
 -    bits = BN_num_bits(DSA_get0_q(dsa_key));
 +    bits = BN_num_bits(dsa_key->q);
    } else if (base_id == EVP_PKEY_EC) {
-     EC_KEY* ec_key = EVP_PKEY_get0_EC_KEY(pkey.get());
+     const EC_KEY* ec_key = EVP_PKEY_get0_EC_KEY(pkey.get());
```
Comment (nornagon) [inline patches/node/fix_use_crypto_impls_for_compat.patch:30]: this bit seems upstreamable maybe?

File: patches/node/make_module_globalpaths_a_reference.patch [hunk 17]
```diff
@@ -7,19 +7,15 @@ We need to hack the search paths of the require function so we can
...
 +++ b/lib/internal/modules/cjs/loader.js
-@@ -1199,8 +1199,8 @@ Module._initPaths = function() {
- 
+@@ -1236,7 +1236,7 @@ Module._initPaths = function() {
    modulePaths = paths;
  
--  // Clone as a shallow copy, for introspection.
--  Module.globalPaths = modulePaths.slice(0);
-+  // clone as a shallow copy, for introspection.
+   // Clone as a shallow copy, for introspection.
```
Comment (nornagon) [inline patches/node/make_module_globalpaths_a_reference.patch:17]:
```suggestion
-  // Clone as a shallow copy, for introspection.
```

File: patches/node/refactor_allow_embedder_overriding_of_internal_fs_calls.patch [hunk 18]
```diff
@@ -60,8 +60,8 @@ index 4a2b0e6ddb3ed8881f896c6f556c9debaf778b81..209df9033a26c29c6e60bd173f710558
      return cache.get(jsonPath);
    }
  
--  const [string, containsKeys] = internalModuleReadJSON(
-+  const [string, containsKeys] = internalFsBinding.internalModuleReadJSON(
+-  const { 0: string, 1: containsKeys } = internalModuleReadJSON(
++  const { 0: string, 1: containsKeys } = internalFsBinding.internalModuleReadJSON(
```
Comment (nornagon) [inline patches/node/refactor_allow_embedder_overriding_of_internal_fs_calls.patch:64]: ‼️,  why is this weird destructuring needed...?

File: patches/node/src_add_get_set_pair_for_unhandled_rejections_mode.patch [hunk 19]
```diff
@@ -0,0 +1,56 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Shelley Vohr <shelley.vohr@gmail.com>
+Date: Thu, 3 Jun 2021 19:32:09 +0200
+Subject: src: add get/set pair for unhandled rejections mode
+
+This PR adds a get/set pair for unhandled rejections modes.
+
+We do not want unhandled rejections to crash the process and want to
```
Comment (nornagon) [inline patches/node/src_add_get_set_pair_for_unhandled_rejections_mode.patch:8]: Do we not? Why do we want to differ from node in this regard?

File: script/node-disabled-tests.json [hunk 20]
```diff
@@ -12,15 +12,22 @@
...
   "parallel/test-crypto-certificate",
   "parallel/test-crypto-des3-wrap",
   "parallel/test-crypto-dh-stateless",
   "parallel/test-crypto-ecb",
   "parallel/test-crypto-engine",
+  "parallel/test-crypto-fips",
+  "parallel/test-crypto-getcipherinfo",
+  "parallel/test-crypto-hkdf.js",
+  "parallel/test-crypto-rsa-dsa",
+  "parallel/test-crypto-secure-heap",
```
Comment (nornagon) [inline script/node-disabled-tests.json:30]: we should be able to enable at least some of these

File: script/node-disabled-tests.json [hunk 21]
```diff
@@ -136,6 +147,19 @@
...
+  "parallel/test-webcrypto-derivebits-node-dh",
+  "parallel/test-webcrypto-ed25519-ed448",
+  "parallel/test-webcrypto-encrypt-decrypt",
+  "parallel/test-webcrypto-encrypt-decrypt-rsa",
+  "parallel/test-webcrypto-rsa-pss-params",
+  "parallel/test-webcrypto-x25519-x448",
+  "parallel/test-webcrypto-wrap-unwrap",
+  "parallel/test-webcrypto-sign-verify-node-dsa",
+  "parallel/test-webcrypto-keygen",
+  "parallel/test-webcrypto-encrypt-decrypt-aes",
```
Comment (nornagon) [inline script/node-disabled-tests.json:162]: hm, none of these work...? :( can we fix them?

File: script/node-disabled-tests.json [hunk 22]
```diff
@@ -34,6 +41,7 @@
   "parallel/test-http2-clean-output",
   "parallel/test-http2-reset-flood",
   "parallel/test-https-agent-session-reuse",
+  "parallel/test-https-client-resume",
```
Comment (nornagon) [inline script/node-disabled-tests.json:44]: why does this break..?

File: spec-main/api-browser-window-spec.ts [hunk 23]
```diff
@@ -3660,11 +3660,12 @@ describe('BrowserWindow module', () => {
...
 
         const didStartLoading = emittedOnce(w.webContents, 'did-start-loading');
-        w.webContents.executeJavaScript(`
+        await w.webContents.executeJavaScript(`(() => {
           var iframe = document.createElement('iframe')
           iframe.src = '${serverUrl}/page2'
           document.body.appendChild(iframe)
-        `);
+          return null
+        })()`);
```
Comment (nornagon) [inline spec-main/api-browser-window-spec.ts:3668]: I don't think the IIFE is needed here, just write `null` at the end of the block.

```suggestion
        await w.webContents.executeJavaScript(`
          var iframe = document.createElement('iframe')
          iframe.src = '${serverUrl}/page2'
          document.body.appendChild(iframe)
          null
        `);
```


... also, should this be split into a separate PR? it seems unrelated.

File: spec-main/extensions-spec.ts [hunk 24]
```diff
@@ -297,30 +298,30 @@ describe('chrome extensions', () => {
...
-          await customSession.loadExtension(path.join(fixtures, 'extensions', 'chrome-webRequest'));
-          fetch(w.webContents, url);
-        })();
-      });
+        await w.loadURL(url);
+        await customSession.loadExtension(path.join(fixtures, 'extensions', 'chrome-webRequest'));
+        fetch(w.webContents, url);
+      };
+
+      await doAsyncWork(run);
```
Comment (nornagon) [inline spec-main/extensions-spec.ts:312]: ... why is this wrapped in a Promise in the first place? can we just remove all these wrappers and write the code straightforwardly?

**Review by nornagon — CHANGES_REQUESTED** [review]

File: patches/boringssl/src_add_impl_for_evp_pkey_get0.patch [hunk 25]
```diff
@@ -0,0 +1,38 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Shelley Vohr <shelley.vohr@gmail.com>
+Date: Mon, 31 May 2021 11:30:38 +0200
+Subject: src: add impl for EVP_PKEY_get0
+
+This adds the missing implementation for EVP_PKEY_get0, which Node.js
+uses and which is present in OpenSSL but not BoringSSL.
+
+This will be upstreamed either fully or as a shim.
```
[the remark this answers is not in the document]
Comment (codebytere) [reply patches/boringssl/src_add_impl_for_evp_pkey_get0.patch:9]: @nornagon already done and merged

File: script/node-disabled-tests.json (hunk 21 above)
Comment (codebytere) [reply script/node-disabled-tests.json:162]: @nornagon i think we should come back to them rather than block this PR on them - that was my plan anyway!

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch [hunk 26]
```diff
@@ -8,86 +8,288 @@ and https://github.com/nodejs/node/pull/31178 that is not currently compatible
...
++#include <openssl/rand.h>
+ 
+ namespace node {
+ 
+@@ -149,7 +150,7 @@ Maybe<bool> RandomPrimeTraits::AdditionalConfig(
+ 
+   params->bits = bits;
+   params->safe = safe;
+-  params->prime.reset(BN_secure_new());
++  params->prime.reset(BN_new());
```
[the remark this answers is not in the document]
Comment (codebytere) [reply patches/node/fix_comment_out_incompatible_crypto_modules.patch:193]: @nornagon it's not supported by BoringSSL at all.

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch (hunk 1 above)
Comment (codebytere) [reply patches/node/fix_comment_out_incompatible_crypto_modules.patch:279]: we didn't, no

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch [hunk 27]
```diff
@@ -8,86 +8,288 @@ and https://github.com/nodejs/node/pull/31178 that is not currently compatible
...
+-
++#ifndef OPENSSL_IS_BORINGSSL
+   if (info->Set(
+           env->context(),
+           env->name_string(),
+           OneByteString(env->isolate(), EVP_CIPHER_name(cipher))).IsNothing()) {
+     return;
+   }
+-
++#endif
```
[the remark this answers is not in the document]
Comment (codebytere) [reply patches/node/fix_comment_out_incompatible_crypto_modules.patch:28]: figured out an alternative!

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch [hunk 28]
```diff
@@ -8,86 +8,288 @@ and https://github.com/nodejs/node/pull/31178 that is not currently compatible
...
+     void* label = OPENSSL_memdup(oaep_label.data(), oaep_label.size());
+     CHECK_NOT_NULL(label);
++#ifndef OPENSSL_IS_BORINGSSL
+     if (0 >= EVP_PKEY_CTX_set0_rsa_oaep_label(ctx.get(),
+                 reinterpret_cast<unsigned char*>(label),
+                                       oaep_label.size())) {
+       OPENSSL_free(label);
+       return false;
+     }
++#endif
```
[the remark this answers is not in the document]
Comment (codebytere) [reply patches/node/fix_comment_out_incompatible_crypto_modules.patch:43]: I'll see what I can do but we previously just commented this same function out in the rsa logic and may not have a good path forward here

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch [hunk 29]
```diff
@@ -8,86 +8,288 @@ and https://github.com/nodejs/node/pull/31178 that is not currently compatible
...
+@@ -210,10 +210,12 @@ WebCryptoCipherStatus RSA_Cipher(
+   if (label_len > 0) {
+     void* label = OPENSSL_memdup(params.label.get(), label_len);
+     CHECK_NOT_NULL(label);
++#ifndef OPENSSL_IS_BORINGSSL
+     if (EVP_PKEY_CTX_set0_rsa_oaep_label(ctx.get(), label, label_len) <= 0) {
+       OPENSSL_free(label);
+       return WebCryptoCipherStatus::FAILED;
+     }
++#endif
```
[the remark this answers is not in the document]
Comment (codebytere) [reply patches/node/fix_comment_out_incompatible_crypto_modules.patch:210]: figured out a fix

File: patches/node/fix_key_gen_apis_are_not_available_in_boringssl.patch (hunk 15 above)
Comment (codebytere) [reply patches/node/fix_key_gen_apis_are_not_available_in_boringssl.patch:93]: probably not but we historically haven't either and i'm not sure how we can do that in this context since nothing in that block is exposed via BoringSSL 🤔

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch [hunk 30]
```diff
@@ -8,86 +8,288 @@ and https://github.com/nodejs/node/pull/31178 that is not currently compatible
...
-       EVP_PKEY_assign_DH(params.get(), dh.release());
+diff --git a/src/crypto/crypto_hkdf.cc b/src/crypto/crypto_hkdf.cc
+index 0aa96ada47abe4b66fb616c665101278bbe0afb6..1e9a4863c5faea5f6b275483ca16f3a6e8dac25b 100644
+--- a/src/crypto/crypto_hkdf.cc
++++ b/src/crypto/crypto_hkdf.cc
+@@ -101,6 +101,7 @@ bool HKDFTraits::DeriveBits(
+     Environment* env,
+     const HKDFConfig& params,
+     ByteSource* out) {
++#ifndef OPENSSL_IS_BORINGSSL
```
[the remark this answers is not in the document]
Comment (codebytere) [reply patches/node/fix_comment_out_incompatible_crypto_modules.patch]: They don't have support for: `EVP_PKEY_CTX_set_hkdf_md`, `EVP_PKEY_CTX_set1_hkdf_salt`, or `EVP_PKEY_CTX_add1_hkdf_info` it doesn't appear - without those I dont think we can really do much here unless i'm perhaps missing something?

File: patches/node/fix_comment_out_incompatible_crypto_modules.patch (hunk 2 above)
Comment (codebytere) [reply patches/node/fix_comment_out_incompatible_crypto_modules.patch]: kept getting this weird error:
```
../../third_party/electron_node/src/node_metadata.h:10:10: fatal error: 'openssl/crypto.h' file not found
#include <openssl/crypto.h>
         ^~~~~~~~~~~~~~~~~~
1 error generated.
```

so i added it as a stopgap before i could look into it further

File: spec-main/extensions-spec.ts (hunk 24 above)
Comment (codebytere) [reply spec-main/extensions-spec.ts]: (see above comment)

File: patches/node/refactor_allow_embedder_overriding_of_internal_fs_calls.patch (hunk 18 above)
Comment (codebytere) [reply patches/node/refactor_allow_embedder_overriding_of_internal_fs_calls.patch:64]: @nornagon this is not our change - it's an update to adapt to [Node.js own code](https://github.com/nodejs/node/blob/5f2bb885f8027e2789a1fa8679d242162b4a8633/lib/internal/modules/package_json_reader.js#L21-L23)

**Review by ckerr — COMMENTED** [review]
Mostly LGTM. This is looking good. +1 on Jeremy's comments about crypto.

Also I lean towards Anna's idea upstream about finding a way to handle unhandled exceptions without adding new undocumented API. Also, I I'm still curious why it's needed, so one nit on this PR might be to update the patch comment to add more background explaining th at

**Comment by codebytere** [comment]
@ckerr @nornagon in v15 Node.js made a change such that unhandled rejections exit the process, which means that any unhandled rejection would now cause the process to exit. I introduced this patch because I think this behavior isn't appropriate for all Electron processes, and that we instead want to selectively enable it. For example, I think that this behavior is fine/expected in `ELECTRON_RUN_AS_NODE`, but that an unhandled rejection shouldn't crash the renderer process.

**Comment by MarshallOfSound** [comment]
> Also I lean towards Anna's idea upstream about finding a way to handle unhandled exceptions without adding new undocumented API

To clarify this, the "undocumented API" is internal to NodeJS in cpp land, there's no public facing API here from Electrons perspective.

> Do we not? Why do we want to differ from node in this regard?

To answer this question though, we already differ from Node in regards to global error handling for precisely the reasons @codebytere outlined above.  What is expected behavior for a NodeJS script / server is entirely unexpected for a desktop application.  One unhandled rejection / exception should not pull the entire app down.

That's actually why we have [this global uncaughtException handler](https://github.com/electron/electron/blob/44491b023ac2653538b7ac36bb73f3085a938666/lib/browser/init.ts#L22-L38) which prevents nodes default behavior of bringing down the process because _by default in electron_ that error is handled.

The approach we took here is slightly different (rather we made the unhandled rejection behavior it's old-form which people expect) but an equivalent solution would be to add a similar global `unhandledRejection` listener such that all rejections are "handled" still.

The TLDR here is expectations of unhandled errors are fundamentally different between desktop apps and node scripts.

**Review by jkleinsc — APPROVED** [review]

File: patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch [hunk 31]
```diff
@@ -0,0 +1,426 @@
...
+-
++#ifndef OPENSSL_IS_BORINGSSL
+   if (!param_ctx ||
+       EVP_PKEY_paramgen_init(param_ctx.get()) <= 0 ||
+       EVP_PKEY_CTX_set_dsa_paramgen_bits(
+@@ -49,7 +49,7 @@ EVPKeyCtxPointer DsaKeyGenTraits::Setup(DsaKeyPairGenConfig* params) {
+       return EVPKeyCtxPointer();
+     }
+   }
+-
```
Comment (nornagon) [inline patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch:212]: do we need a `#else` `return` here?

File: patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch [hunk 32]
```diff
@@ -0,0 +1,426 @@
...
+--- a/src/crypto/crypto_rsa.cc
++++ b/src/crypto/crypto_rsa.cc
+@@ -210,7 +210,10 @@ WebCryptoCipherStatus RSA_Cipher(
+   if (label_len > 0) {
+     void* label = OPENSSL_memdup(params.label.get(), label_len);
+     CHECK_NOT_NULL(label);
+-    if (EVP_PKEY_CTX_set0_rsa_oaep_label(ctx.get(), label, label_len) <= 0) {
++    if (EVP_PKEY_CTX_set0_rsa_oaep_label(
++      ctx.get(),
++      reinterpret_cast<unsigned char*>(label),
```
Comment (nornagon) [inline patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch:284]: nit: the underlying type in boringssl here is `uint8_t`, so perhaps
```suggestion
+      reinterpret_cast<uint8_t*>(label),
```

File: shell/app/node_main.cc [hunk 33]
```diff
@@ -74,6 +74,7 @@ void SetNodeCliFlags() {
     if (disallowed.count(stripped) != 0) {
       LOG(ERROR) << "The Node.js cli flag " << stripped
                  << " is not supported in Electron";
+      return 9;
```
Comment (nornagon) [inline shell/app/node_main.cc:77]: Why 9? Is this documented anywhere?

**Review by nornagon — APPROVED** [review]
lgtm overall, v happy to see all the upstreamed crypto stuff!

File: patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch (hunk 32 above)
Comment (codebytere) [reply patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch:284]: I upstreamed it [here](https://github.com/nodejs/node/pull/38926) and Node.js uses `static_cast` by convention so let's switch to that since it's what we'll eventually receive from upstream?

File: patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch (hunk 31 above)
Comment (codebytere) [reply patches/node/fix_handle_boringssl_and_openssl_incompatibilities.patch:221]: ah yeah we should!

**Review by ckerr — APPROVED** [review]
